### DISCUSSION TITLE
code style, *const T

### DISCUSSION THREAD

**Discussion by iacore** [david-vanderson/dvui#54] [General]
I see a lot of `self: *const Options` in the code. Because of Zig's [janky result location semantics](https://github.com/ziglang/zig/issues/12064), I think passing `*const Options` is the same as passing `Options`.

example buggy code by Lemon Drop

```
const AAAA = struct {
    foo: [100]u32,
};

fn aaaaa(a: AAAA, b: *AAAA) void {
  b.*.foo[0] = 5;

  std.debug.print("Wtf: {}", .{ a.foo[0] });
}

pub fn main() !void {
    var f: AAAA = undefined;

    f.foo[0] = 0;

    aaaaa(f, &f);
}
```

**Comment by david-vanderson** [comment]
Read my story here - https://github.com/ziglang/zig/issues/13249

I got burned at least twice by code that thought it was operating on "the object we called the function on", but was really operating on a copy.  So now I always make `self` a pointer argument.  I've heard other projects (tigerbeetle?) do the same.

**Reply by iacore** [reply]
I don't think Zig takes a copy when you pass struct as parameter

```
var s = S{};
foo(s); // s.foo()
```

It may take a copy, or it may pass by reference.

See https://ziglang.org/documentation/0.11.0/#Pass-by-value-Parameters

**Reply by david-vanderson** [reply]
Right - it is up to the compiler whether to pass a copy or pass a const pointer to the original.  That choice matters if you take a pointer to it or a field of it inside the function and return it outside.  This doesn't happen that often, but like I said I've been bitten at least twice already.  Spelling it out as a pointer means I don't have to think about it.

I'm not going to insist that anybody else write code that way, that's just an explanation for why I write it that way.  Does that make sense?

**Reply by iacore** [reply]
yes that makes sense
